Importing the dependencies:

In [48]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [49]:
#loading the dataset into the pandas dataframe
credit_card_df = pd.read_csv('/content/drive/MyDrive/creditcard.csv')

In [50]:
#checking the first five rows
credit_card_df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [51]:
#checking the last five rows
credit_card_df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

In [52]:
#dataset information
credit_card_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [53]:
#checking the number of missing values in each column
credit_card_df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

As we can see there is no missing values at all in the dataset. So it is plus point for us. Else we will have to handle the missing values by dropping them or by replacing them with some other values, this method is know as imputation.

In [54]:
credit_card_df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

As we can see the number of values of two cases in our target variable is too much different which states that our dataset is too imbalanced.So, we will have to handle this by using the method such as undersampling where we will take some random sample from our data and will try to make it balanced.So that our model can be trained in a generalise way.

0 --> Normal Transaction

1 --> Fraudlent Transaction

In [55]:
#separating the data for the analysis
legit = credit_card_df[credit_card_df.Class == 0]
fraud = credit_card_df[credit_card_df.Class == 1]

In [56]:
legit.shape

(284315, 31)

In [57]:
fraud.shape

(492, 31)

In [58]:
#statistical measure about the data of a particular column
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [59]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [60]:
#comparing the values of both sort of transactions using groupby
credit_card_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Under-Sampling: As the dataset is highly imbalanced so that is a technique to overcome this. In this, we will build sample dataset containing similar distribution of normal and fraudlent transactions.

As the number of fraudlent transactions is 492, so we will build a sample dataset of normal transactions too with 492 transactions only.

In [61]:
legit_sample = legit.sample(n=492)

In [62]:
legit_sample.shape

(492, 31)

In [63]:
#concatenating the two datasets
new_df = pd.concat([legit_sample, fraud], axis = 0)

In [64]:
new_df.shape

(984, 31)

In [65]:
new_df.head()

Time        V1        V2        V3        V4        V5        V6  \
162136  114858.0  2.004613  0.055839 -1.621582  0.356083  0.333140 -0.672122   
159608  112716.0  0.109095  0.820418 -0.149979 -0.825409  0.893693 -0.524624   
202500  134357.0  1.757293 -0.724475 -1.606448  0.123006  0.232973 -0.158640   
87241    61597.0  1.255339  0.223388  0.266252  0.709790 -0.530448 -1.113199   
268008  163024.0  1.993492 -0.376313 -0.348721  0.334432 -0.792799 -0.757987   

              V7        V8        V9  ...       V21       V22       V23  \
162136  0.064450 -0.076762  0.291471  ... -0.301606 -0.839825  0.349767   
159608  1.002682 -0.138959  0.034570  ... -0.326690 -0.768341 -0.046445   
202500  0.285785 -0.257542  0.910179  ... -0.094312 -0.386933 -0.007808   
87241  -0.035945 -0.100835  0.306492  ... -0.308733 -0.978911  0.150377   
268008 -0.568603  0.017255  1.574817  ... -0.136120 -0.322994  0.371336   

             V24       V25       V26       V27       V28  Amount  Class  
162136  0.657358 -0.331706  0.143373 -0.070694 -0.038476    8.99      0  
159608 -1.012910 -0.350652  0.202805  0.243677  0.087102    8.08      0  
202500  0.142584  0.025892 -0.105163 -0.055480 -0.021905  183.56      0  
87241   0.290336  0.155664  0.100802 -0.034894  0.026888    1.79      0  
268008 -0.187653 -0.427881 -0.911162  0.051876 -0.031685    1.00      0  

[5 rows x 31 columns]

In [66]:
new_df.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

All the fraudulent transactions are concatenated on the top of normal transactions, which can be seen using tail function.

In [67]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

Now we can see, our dataset is fully balanced.

In [68]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      99273.396341 -0.162917 -0.245653  0.007081  0.136221  0.112538   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.118362  0.045911 -0.173654  0.002961  ...  0.020737 -0.062647   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.041032  0.036359  0.013893 -0.004935 -0.009740 -0.059724  0.052754   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0      107.255407  
1      122.211321  

[2 rows x 30 columns]

Now we can see by grouping the transaction by mean that there is no huge difference in between the values of normal and fraudulent transactions.

Splitting the data into features and targets:

In [69]:
X = new_df.drop(columns = 'Class', axis = 1)
Y = new_df['Class']

In [70]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
162136  114858.0  2.004613  0.055839 -1.621582  0.356083  0.333140 -0.672122   
159608  112716.0  0.109095  0.820418 -0.149979 -0.825409  0.893693 -0.524624   
202500  134357.0  1.757293 -0.724475 -1.606448  0.123006  0.232973 -0.158640   
87241    61597.0  1.255339  0.223388  0.266252  0.709790 -0.530448 -1.113199   
268008  163024.0  1.993492 -0.376313 -0.348721  0.334432 -0.792799 -0.757987   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [71]:
print(Y)

162136    0
159608    0
202500    0
87241     0
268008    0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Splitting the data into training data and testing data for training the model.

In [72]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, stratify=Y, random_state = 2)

In [73]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


Model training using Logistic Regression

In [74]:
model = LogisticRegression()

In [75]:
#training the model with training data
model.fit(X_train, Y_train)

LogisticRegression()

Model building on imbalanced data:


Metric selection for heavily imbalanced data As we have seen that the data is heavily imbalanced, where only 0.17% transctions are fraudulent, we should not consider Accuracy as a good measure for evaluating the model. Because in the case of all the datapoints return a particular class(1/0) irrespective of any prediction, still the model will result more than 99% Accuracy.


Hence, we have to measure the ROC-AUC score for fair evaluation of the model. The ROC curve is used to understand the strength of the model by evaluating the performance of the model at all the classification thresholds. The default threshold of 0.5 is not always the ideal threshold to find the best classification label of the test point. Because the ROC curve is measured at all thresholds, the best threshold would be one at which the TPR is high and FPR is low, i.e., misclassifications are low. After determining the optimal threshold, we can calculate the F1 score of the classifier to measure the precision and recall at the selected threshold.


Why SVM was not tried for model building and Random Forest was not tried for few cases?


In the dataset we have 284807 datapoints and in the case of Oversampling we would have even more number of datapoints. SVM is not very efficient with large number of datapoints beacuse it takes lot of computational power and resources to make the transformation. When we perform the cross validation with K-Fold for hyperparameter tuning, it takes lot of computational resources and it is very time consuming. Hence, because of the unavailablity of the required resources and time SVM was not tried.


For the same reason Random forest was also not tried for model building in few of the hyperparameter tuning for oversampling technique.


Why KNN was not used for model building?


KNN is not memory efficient. It becomes very slow as the number of datapoints increases as the model needs to store all the data points. It is computationally heavy because for a single datapoint the algorithm has to calculate the distance of all the datapoints and find the nearest neighbors.


Model Evaluation:

Accuracy Score:




In [76]:
#accuracy on the training data
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [77]:
print('Accuracy on the training data:', training_data_accuracy)

Accuracy on the training data: 0.9301143583227446


In [82]:
#accuracy on the test data
X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [83]:
print('Accuracy on the test data:', test_data_accuracy )

Accuracy on the test data: 0.8984771573604061
